workspaces/optic-engine/src/projections/spectacle/endpoints.rs

```diff
@@ -529,11 +527,65 @@ impl EndpointsProjection {
       .domain_id_to_index
       .get(&request_id)
       .expect("expected node with domain_id $request_id to exist in the graph");
+
+    let endpoint_id = self.get_endpoint_id_from_node(&request_id);
+    let endpoint_index = *self
+      .domain_id_to_index
+      .get(&endpoint_id)
+      .expect("expected node to exist in graph");
+
     let node = Node::Body(BodyNode {
-      http_content_type: http_content_type,
+      http_content_type: http_content_type.clone(),
       root_shape_id: root_shape_id.clone(),
       is_removed: false,
     });
+
+    // If there is already a set request body with a matching content type, we should orphan the old node
+    // @GOTCHA - this only allows a single content type request parameter to be attached to an endpoint node
+    // This means history is only visible to users for the latest request node with content type
+    // @GOTCHA - event generation does not group together same path + method, meaning we need to search
+    // all request ids attached to an endpoint node (i.e. request -> body is 1:1 since request nodes are never reused)
```

Comment:
So if we want to fix this, I think we will need to change the event generation such that adding a new request looks for existing requests. Right now requests and responses are newly created without context of the current spec, so a request node is always created. I'm not sure where this is created (likely both the UI and rust code).

We'll also probably need to figure out some migration path / handling old event shapes since this code expects a single body per request / response node (how events are currently generated)

Reply:
> Right now requests and responses are newly created without context of the current spec, so a request node is always created.

Happens in the undocumented bodies learner in the Engine, giving the UI the commands it can apply. Not sure if the UI mutates those in any way (doubt it / hope not).